src/scitacean/transfer/ssh.py

```diff
@@ -73,50 +82,99 @@ def download_file(self, *, remote: str, local: Path):
         self._connection.get(remote=str(remote), local=str(local))
 
 
-class _ESSUploadConnection:
+class SSHUploadConnection:
     def __init__(
         self, *, connection: Connection, dataset_id: PID, remote_base_path: str
     ):
         self._connection = connection
         self._dataset_id = dataset_id
         self._remote_base_path = remote_base_path
 
+    @property
+    def _sftp(self) -> SFTPClient:
+        return self._connection.sftp()
+
     @property
     def source_dir(self) -> str:
         return os.path.join(self._remote_base_path, self._dataset_id.pid)
 
     def remote_path(self, filename) -> str:
         return os.path.join(self.source_dir, filename)
 
-    def upload_files(self, *files: File) -> List[File]:
-        """Upload files to the remote staging folder."""
-        return [
-            file.uploaded(
-                remote_path=self.upload_file(
-                    remote=file.remote_path, local=file.local_path
-                )
-            )
-            for file in files
-        ]
+    def _make_source_folder(self):
+        try:
+            self._sftp.mkdir(self.source_dir)
+        except OSError as exc:
+            raise FileUploadError(
+                f"Failed to create source folder: {exc.args}"
+            ) from None
 
-    def upload_file(self, *, remote: Union[str, Path], local: Union[str, Path]) -> str:
-        """Upload a file to the remote staging folder."""
-        remote_path = self.remote_path(remote)
+    def upload_files(self, *files: File) -> List[File]:
+        """Upload files to the remote folder."""
+        self._make_source_folder()
+        uploaded = []
+        try:
+            for file in files:
+                uploaded.append(self._upload_file(file))
+        except Exception:
+            self.revert_upload(*uploaded)
```

Comment:
Continue in what way? Upload a changed dataset with fewer files?
Also keep in mind that this can happen if the uploaded file is broken (e.g. fails the checksum test). In this case, we don't want to make a dataset. In particular because files will we read only once they are properly ingested (something DST will have to implement)

Reply:
I was more thinking of leaving the uploaded files, and when you try again, compare size and checksum (and do nothing if they match)?

Reply:
This is a bit tricky for non-expert users. The client assigns a new ID to the dataset during upload (to avoid clashes in the database and filesystem and stop people from doing shenanigans). The SSHFileTransfer then uses the ID to create a new unique directory and place the files in that.
When we call the upload function again with the same input, it will assign a new ID and therefore upload to a different folder. Since the dataset from the first attempt never made it into SciCat, we lost its ID (unless we log it and the user happens to save the log at the appropriate level). So we cannot even (reliably) find the files anymore.

I agree that experienced users can fix the files, they can, e.g. catch the ID form log messages and assign it to the dataset themselves, or see the logs for the file uploads. Or we put the path into the exception message. But all of these are easy to lose.

In addition, it is not clear yet how the file server will behave. Files will ultimately become read only after they have been created. So it is possible that they cannot be fixed or even reverted. Though I hope that at least the latter will be possible.

Reply:
Hmm, would it make more sense to *first* upload the files and *then* create the dataset?

Reply:
That is what it does. But it needs to know where to put the files. And guarantee that it doesn't clash with any other dataset's files. You could of course do this manually but then it is up to the scientists to handle their files correctly

Reply:
👍